src/xenia/base/clock.cc

```diff
@@ -10,50 +10,136 @@
 #include "xenia/base/clock.h"
 
 #include <algorithm>
-#include <climits>
+#include <limits>
+#include <mutex>
 
 #include "xenia/base/assert.h"
+#include "xenia/base/math.h"
+
+DEFINE_bool(clock_no_scaling, false,
+            "Disable scaling code. Time management and locking is bypassed. "
+            "Guest system time is directly pulled from host.",
+            "CPU");
+DEFINE_bool(clock_source_raw, false,
+            "Use the RDTSC instruction as the time source. "
+            "Host CPU must support invariant TSC. ",
+            "CPU");
 
 namespace xe {
 
 // Time scalar applied to all time operations.
 double guest_time_scalar_ = 1.0;
 // Tick frequency of guest.
-uint64_t guest_tick_frequency_ = Clock::host_tick_frequency();
+uint64_t guest_tick_frequency_ = Clock::host_tick_frequency_platform();
 // Base FILETIME of the guest system from app start.
 uint64_t guest_system_time_base_ = Clock::QueryHostSystemTime();
-// Combined time and frequency scalar (computed by RecomputeGuestTickScalar).
-double guest_tick_scalar_ = 1.0;
+// Combined time and frequency ratio between host and guest.
+// Split in numerator (first) and denominator (second).
+// Computed by RecomputeGuestTickScalar.
+std::pair<uint64_t, uint64_t> guest_tick_ratio_ = std::make_pair(1, 1);
+
 // Native guest ticks.
-thread_local uint64_t guest_tick_count_ = 0;
-// 100ns ticks, relative to guest_system_time_base_.
-thread_local uint64_t guest_time_filetime_ = 0;
+uint64_t last_guest_tick_count_ = 0;
 // Last sampled host tick count.
-thread_local uint64_t last_host_tick_count_ = Clock::QueryHostTickCount();
+uint64_t last_host_tick_count_ = Clock::QueryHostTickCount();
+// Mutex to ensure last_host_tick_count_ and last_guest_tick_count_ are in sync
+std::mutex tick_mutex_;
 
 void RecomputeGuestTickScalar() {
-  guest_tick_scalar_ = (guest_tick_frequency_ * guest_time_scalar_) /
-                       static_cast<double>(Clock::host_tick_frequency());
+  // Create a rational number with numerator (first) and denominator (second)
+  auto frac =
+      std::make_pair(guest_tick_frequency_, Clock::QueryHostTickFrequency());
+  // Doing it this way ensures we don't mess up our frequency scaling and
+  // precisely controls the precision the guest_time_scalar_ can have.
+  if (guest_time_scalar_ > 1.0) {
+    frac.first *= static_cast<uint64_t>(guest_time_scalar_ * 10.0);
+    frac.second *= 10;
+  } else {
+    frac.first *= 10;
+    frac.second *= static_cast<uint64_t>(10.0 / guest_time_scalar_);
+  }
+  // Keep this a rational calculation and reduce the fraction
+  reduce_fraction(frac);
+
+  std::unique_lock<std::mutex> lock(tick_mutex_);
```

Comment:
```suggestion
  std::lock_guard lock(tick_mutex_);
```

`std::lock_guard<std::mutex>` if we don't have C++17 on all platforms.